### PR TITLE
co-processor: add snapshot and rpc handler

### PR DESCRIPTION
**Pull request by disksing**

### CODE REVIEW

File: src/server/coprocessor.rs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    }
+
+    #[allow(unused_variables)]
+    pub fn on_request(&self, req: Request, token: Token, msg_id: u64) -> Result<()> {
+        unimplemented!();
+    }
+
+    #[allow(dead_code)]
+    fn new_snapshot<'a>(&'a self, ctx: &Context, start_ts: u64) -> Result<SnapshotStore<'a>> {
+        let snapshot = try!(self.engine.get_snapshot(ctx).map_err(StorageError::from));
```
Comment (BusyJay) [inline src/server/coprocessor.rs]: Why not just use try?

File: src/server/server.rs [hunk 2]
```diff
@@ -199,6 +203,7 @@ impl<T: RaftStoreRouter> Server<T> {
             }
             MessageType::Cmd => self.on_raft_command(msg.take_cmd_req(), token, msg_id),
             MessageType::KvReq => self.store.on_request(msg.take_kv_req(), token, msg_id),
+            MessageType::CopReq => self.coprocessor.on_request(msg.take_cop_req(), token, msg_id),
```
Comment (BusyJay) [inline src/server/server.rs]: coprocessor may take a long time to finish.

File: src/storage/mod.rs [hunk 3]
```diff
@@ -159,14 +161,16 @@ impl fmt::Debug for Command {
...
+    engine: Arc<Box<Engine>>,
     tx: Sender<Message>,
     thread: JoinHandle<Result<()>>,
 }
 
 impl Storage {
     pub fn from_engine(engine: Box<Engine>) -> Result<Storage> {
         let desc = format!("{:?}", engine);
-        let mut scheduler = Scheduler::new(engine);
+        let shared = Arc::new(engine);
```
Comment (BusyJay) [inline src/storage/mod.rs]: Why named shared?

**Comment by BusyJay** [comment]
rest LGTM

File: src/server/server.rs (hunk 2 above)
Comment (disksing) [inline src/server/server.rs]: Yeah. We will add thread inside coprocessor (just like Storage) later, this PR concentrates on the snapshot.

Comment (BusyJay) [inline src/server/server.rs]: Then we should add a todo.

File: src/storage/mod.rs (hunk 3 above)
Comment (disksing) [inline src/storage/mod.rs]: Arc is used for sharing the engine between threads here. The name 'shared' is inspired by the Arc document: https://doc.rust-lang.org/stable/std/sync/struct.Arc.html#examples

File: src/server/coprocessor.rs [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+use kvproto::coprocessor::Request;
+use super::{Result, SendCh};
+
+#[allow(dead_code)]
+pub struct CoprocessorHandler {
+    engine: Arc<Box<Engine>>,
+    ch: SendCh,
+}
+
+impl CoprocessorHandler {
```
Comment (ngaut) [inline src/server/coprocessor.rs]: s/CoprocessorHandler/SnapshotCorprocessor

**Comment by BusyJay** [comment]
LGTM

**Comment by BusyJay** [comment]
Ping

File: src/storage/engine/mod.rs [hunk 5]
```diff
@@ -33,6 +33,7 @@ pub trait Engine: Send + Sync + Debug {
     fn get(&self, ctx: &Context, key: &Key) -> Result<Option<Value>>;
     fn seek(&self, ctx: &Context, key: &Key) -> Result<Option<KvPair>>;
     fn write(&self, ctx: &Context, batch: Vec<Modify>) -> Result<()>;
+    fn get_snapshot<'a>(&'a self, ctx: &Context) -> Result<Box<Snapshot + 'a>>;
```
Comment (BusyJay) [inline src/storage/engine/mod.rs]: `snapshot` is enough.

File: src/server/coprocessor.rs [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+use super::{Result, SendCh};
+
+#[allow(dead_code)]
+pub struct SnapshotCoprocessor {
+    engine: Arc<Box<Engine>>,
+    ch: SendCh,
+}
+
+impl SnapshotCoprocessor {
+    pub fn new(engine: Arc<Box<Engine>>, ch: SendCh) -> SnapshotCoprocessor {
```
Comment (siddontang) [inline src/server/coprocessor.rs]: can we use raftstore here directly?

Comment (disksing) [inline src/server/coprocessor.rs]: No, tikv may use other Engines (memory, rocksdb, etc.)

**Comment by BusyJay** [comment]
LGTM

File: src/storage/mvcc/txn.rs [hunk 7]
```diff
@@ -195,6 +195,61 @@ impl<'a, T: Engine + ?Sized> MvccTxn<'a, T> {
...
+                // There is a pending lock. Client should wait or clean it.
+                return Err(Error::KeyIsLocked {
+                    key: key.raw().to_owned(),
+                    primary: lock.get_primary_key().to_vec(),
+                    ts: lock.get_start_ts(),
+                });
+            }
+        }
+        // Find the latest write below our start timestamp.
+        match meta.iter_items().find(|x| x.get_commit_ts() <= ts) {
```
Comment (siddontang) [inline src/storage/mvcc/txn.rs]: binary_search may better? because the meta items are sorted.

**Comment by disksing** [comment]
PTAL @siddontang @ngaut

**Comment by siddontang** [comment]
LGTM

**Comment by ngaut** [comment]
LGTM

File: src/storage/mvcc/txn.rs (hunk 7 above)
Comment (disksing) [inline src/storage/mvcc/txn.rs]: No need. Since timestamps are stored in desc order. We are able to get the version we want for a few iterations.

**Pull request merged**